File: src/game/unit/Unit.cpp
//
// Created by Per-Arne on 24.02.2017.
//

#include "Unit.h"
#include "../player/Player.h"
#include "../Game.h"
#include "../util/Pathfinder.h"
#include "../graphics/GUI.h"
#include "../action/RightClickAction.h"
#include "./UnitManager.h"
#include <random>
#include "../action/BaseAction.h"

int Unit::gId = 0;
Unit::Unit(Player *player): player_(player), stateManager(&player->game_.stateManager){
    id = player->game_.units.size();
    state = stateManager->despawnedState;


}
void Unit::spawn(Tile &_toSpawnOn, int initValue) {
    spawnTimer = initValue;
    spawnTileID = _toSpawnOn.id;
    transitionState(stateManager->spawnState);
	enqueueState(stateManager->idleState);
}

void Unit::moveRelative(int x, int y) {


    int newX = tile->x + x;
    int newY = tile->y + y;

    Tile *moveTile = player_->game_.map.getTile(newX, newY);

    move(*moveTile);
}

void Unit::rightClickRelative(int x, int y) {
	if (!tile) return; // Not standing on any tiles

    int newX = tile->x + x;
    int newY = tile->y + y;

    Tile *clickTile = player_->game_.map.getTile(newX, newY);
    rightClick(*clickTile);
}

void Unit::move(Tile &targetTile){
    if (!canMove)
        return;
    this->walkingGoalID = targetTile.id;
    transitionState(stateManager->walkingState);
}


void Unit::setPosition(Tile &newTile) {

    if(tile){                       // If unit already stands on a tile
        clearTiles();
    }

    for(auto &t : player_->game_.map.getTiles(&newTile, width, height)) {
        t->setOccupant(this);
    }


    newTile.setOccupant(this);         // Set occupant of new tile to this
    tile = &newTile;                     // Set this units tile to new tile
    Position newPos = tile->getPosition();
    setDirection(newPos.x, newPos.y);
    worldPosition = newPos;
}

void Unit::update() {
    state->update(*this);

}

Tile *Unit::centerTile() {
	int addX = floor(width / 2);
	int addY = floor(height / 2);

	if (addX == 0 && addY == 0) {
		assert(tile);
		return tile;
	}

	return player_->game_.map.getTile(tile->x + addX, tile->y + addY);
}

bool Unit::build(int idx) {
    //if(state->id != Constants::State::Idle)
    if(state->id != Constants::State::Idle)
        return false;

    if((idx < 0 or idx >= buildInventory.size()))
        return false;

	Unit newUnit = UnitManager::constructUnit(buildInventory[idx], player_);

    // PlacementTile is based on dimension of the new unit. For example; town hall has
    // 3x Width and 3x Height. We then want to place  the building by the middle tile;
    Tile *placementTile = player_->game_.map.getTile(tile->x - floor(newUnit.width/2), tile->y - floor(newUnit.height/2));
	assert(placementTile && "PlacementTile was null in build()");
    if(!player_->canAfford(newUnit)) {
        //std::cout << "Cannot afford " << newUnit->name << std::endl;
        return false;
    }



    if(player_->canPlace(*this, newUnit, placementTile)) {

        Unit &unit = player_->addUnit(newUnit.typeId);
		unit.player_ = player_;
		unit.builtByID = this->id;

        if(!structure and unit.structure) {
           // *this is a unit (peasant), which builds a building
            despawn();

            buildEntityID = unit.id; // Set id of buildEntity
            transitionState(stateManager->buildingState);
            unit.spawn(*placementTile, 0);     // Spawn build entity
            unit.setPosition(*placementTile);  // Set position of build entity


        }else if(structure and !unit.structure){
            // Structure builds unit (Ie: TownHall builds Peasant)
            buildEntityID = unit.id;    // Set build entity ID
            // build entity has no tile, spawn on nearest walkable
            Tile *firstWalkable = Pathfinder::find_first_walkable_tile(centerTile());
            assert(firstWalkable);
            unit.spawn(*firstWalkable, 0);  // Spawn build entity

            transitionState(stateManager->buildingState);
        }


        player_->subGold(unit.goldCost);
        player_->subLumber(unit.lumberCost);
        player_->subOil(unit.oilCost);
        player_->statUnitBuilt += 1;

        return true;

    } else {
        return false;
    }


    return false;
}

void Unit::despawn() {

    for(auto &p: player_->game_.players) {
        if(p.targetedUnitID == this->id)
            p.targetedUnitID = -1;
    }

    clearTiles();
    transitionState(stateManager->despawnedState);
}

void Unit::clearTiles(){
    for(auto &t : player_->game_.map.getTiles(tile, width, height)) {
        t->setOccupant(NULL);
    }
    tile->setOccupant(NULL);
    tile = NULL;
}

void Unit::rightClick(Tile &tile) {
	stateList.clear();
	transitionState();


    std::shared_ptr<BaseAction> action = std::shared_ptr<BaseAction>(new RightClickAction(*this));
    player_->game_.addAction(action);


    if(tile.isHarvestable()){
        //  Harvest
        harvest(tile);
    }
    else if(tile.isAttackable(*this)){
        // Attack
        attack(tile);
    }
    else if(tile.isWalkable()){
        // Walk
        move(tile);
    }
}

void Unit::attack(Tile &tile) {
    if(!canAttack)
        return;

    Unit* target = tile.getOccupant();

	// Target may have died from another same tick
	if (!target) {
		transitionState();
		return;
	}

    combatTargetID = target->id;

    if(distance(tile) > 1){
        enqueueState(stateManager->combatState);
        move(tile);
    }
    else {
        transitionState(stateManager->combatState);

    }



}

void Unit::harvest(Tile &tile) {
    if(!canHarvest)
        return;

    this->harvestTargetID = tile.id;
    transitionState(stateManager->harvestingState);

}

void Unit::enqueueState(std::shared_ptr<BaseState> state) {
    stateList.push_back(state);
}

void Unit::transitionState() {
    if(stateList.empty()) {
        // No states to transition to, enter idle state
		std::shared_ptr<BaseState> nextState = stateManager->idleState;
        //std::cout << "State Transition: " << state->name << " ==> " << nextState->name << "|" << std::endl;
		//std::cout << "State Transition: " << state->id << " ==> " << nextState->id << "|" << std::endl;
		state->end(*this);
        state = nextState;
        state->init(*this);
        return;
    }

	std::shared_ptr<BaseState> nextState = stateList.back();
    //std::cout << "State Transition: " << state->name << " ==> " << nextState->name << "|y" << std::endl;
	//std::cout << "State Transition: " << state->id << " ==> " << nextState->id << "|y"  << std::endl;
	state->end(*this);
    state = nextState;
    stateList.pop_back();
    //state->init(*this);


}

void Unit::transitionState(std::shared_ptr<BaseState> nextState) {
    //std::cout << "State Transition: " << state->name << " ==> " << nextState->name << "|x" << std::endl;
	//std::cout << "State Transition: " << state->id << " ==> " << nextState->id << "|x" << std::endl;
	state->end(*this);
    state = nextState;
    state->init(*this);
    return;
}

int Unit::distance(Tile &target) {
    int dim = 0; // TODO
    double d = hypot(tile->x - (target.x + dim), tile->y - (target.y + dim));
    return (int)d - dim;

}


int Unit::distance(Unit & target) {
	int targ_x = target.tile->x;
	int targ_y = target.tile->y;
	int dim_x = floor(target.width / 2);
	int dim_y = floor(target.height / 2);

	double d = hypot(tile->x - (targ_x + dim_x), tile->y - (targ_y + dim_y));

	return (int)d - dim_x;

}

Position Unit::distanceVector(Tile &target){
    int dx = tile->x - target.x;
    int dy = tile->y - target.y;

    return Position(dx, dy);
}

Unit* Unit::closestRecallBuilding() {
    Unit* closest = NULL;
    int dist = INT_MAX;
    for(auto &unit : player_->game_.units) {
        if(unit.recallable && unit.player_ == player_ && unit.tile) {
            int d = distance(*unit.tile);
            if(d < dist) {
                dist = d;
                closest = &unit;
            }
        }
    }
    return closest;
}

void Unit::afflictDamage(int dmg_) {
    health = std::max(0, health - dmg_);

    if (health <= 0) {
        transitionState(stateManager->deadState);

        player_->checkDefeat();
        player_->game_.checkTerminal();
    }

}

bool Unit::isDead() {
    return state->id == Constants::State::Dead or state->id == Constants::State::Despawned;
}

int Unit::getDamage(Unit &target) {

    // TODO better random
    double output = damageMin + (rand() % (damageMax - damageMin + 1));
    double myDamage = output - target.armor + damagePiercing;

    double mini = myDamage * .50;
    double output2 = mini + (rand() % (int)(myDamage - mini + 1));

    return floor(output2);

}

void Unit::setDirection(Position &dir){
    setDirection(dir.x, dir.y);
}

void Unit::setDirection(int newX, int newY){
    int oldX = worldPosition.x;
    int oldY = worldPosition.y;

    int dx = (newX - oldX);
    int dy = (newY - oldY);

    if (dx > 0 and dy > 0) {
        // Down Right
        direction = Constants::Direction::DownRight;
        //std::cout << "Down Right" << std::endl;
    } else if (dx < 0 and dy > 0) {
        // Down Left
        direction = Constants::Direction::DownLeft;
        //std::cout << "Down Left" << std::endl;
    } else if (dx > 0 and dy < 0) {
        // Up Right
        direction = Constants::Direction::UpRight;
        //std::cout << "Up Right" << std::endl;
    } else if (dx < 0 and dy < 0) {
        // Up Left
        direction = Constants::Direction::UpLeft;
        //std::cout << "Up Left" << std::endl;
    } else if (dx > 0 and dy == 0) {
        // Right
        direction = Constants::Direction::Right;
        //std::cout << "Right" << std::endl;
    } else if (dx < 0 and dy == 0) {
        // Left
        direction = Constants::Direction::Left;
        //std::cout << "Left" << std::endl;
    } else if (dx == 0 and dy < 0) {
        // Up
        direction = Constants::Direction::Up;
        //std::cout << "Up" << std::endl;
    } else if (dx == 0 and dy > 0) {
        // Down
        direction = Constants::Direction::Down;
        //std::cout << "Down" << std::endl;
    }



}

bool Unit::operator==(int otherID) const
{
	return otherID == id;
}

void Unit::tryAttack()
{
	if (!tile) {
		// FAIL
		return;
	}

	std::vector<Tile *> availableAttackable = player_->game_.getMap().neighbors(*tile, Constants::Pathfinding::Attackable);
	if (availableAttackable.empty()) {
		// Fail
		return;
	}else {
		// Success
		attack(*availableAttackable.back());
	}

}

void Unit::tryMove(int16_t x, int16_t y)
{
	if (!tile) {
		// FAil
		return;
	}

	int newX = tile->x + x;
	int newY = tile->y + y;

	Tile *tile = player_->game_.map.getTile(newX, newY);
	if (tile->isWalkable()) {
		move(*tile);
	}
	else {
		// Fail
		return;
	}
}

void Unit::tryHarvest()
{
	if (!tile) {
		// FAIL
		return;
	}

	std::vector<Tile *> availableHarvestable = player_->game_.getMap().neighbors(*tile, Constants::Pathfinding::Harvestable);
	if (availableHarvestable.empty()) {
		// Fail
		return;
	}
	else {
		harvest(*availableHarvestable.back());
	}
}


Tile &Unit::getSpawnTile() {
    assert(spawnTileID != -1);
    return player_->game_.map.tiles[spawnTileID];
}

Tile *Unit::getTile(int tileID) {
    if(tileID == -1) {
        return NULL;
    }
    return &player_->game_.map.tiles[tileID];
}

Unit &Unit::getBuiltBy() {
    assert(builtByID != -1);
    return player_->game_.units[builtByID];
}

Unit &Unit::getBuildEntity() {
    assert(buildEntityID != -1);
    return player_->game_.units[buildEntityID];
}

Unit &Unit::getCombatTarget() {
    assert(combatTargetID != -1);
    return player_->game_.units[combatTargetID];
}

File: python/game/unit/Unit.py
import itertools
import random
import math
from collections import deque

from game import Config
from game import Constants
from game.unit.UnitManager import UnitManager
from game.util.Pathfinding import a_star_search


class Unit:

    def __init__(self, player):
        self.player = player
        self.id = len(player.game.units)
        self.state_manager = player.game.state_manager
        self.state = player.game.state_manager.despawned_state

        self.animation_interval = .5 * Config.TICK_MULTIPLIER
        self.animation_timer = 0
        self.animation_iterator = 0

        self.type_id = Constants.Unit.Nothing
        self.health = None
        self.health_max = None
        self.direction = None

        self.ground_unit = None
        self.water_unit = None

        self.damage_min = -1
        self.damage_max = -1
        self.damage_range = -1
        self.damage_piercing = -1
        self.armor = -1

        self.lumber_carry = 0
        self.gold_carry = 0
        self.oil_carry = 0
        self.carry_capacity = 10
        self.build_inventory = None
        self.speed = None
        self.sight = None
        self.range = 4

        self.can_harvest = None
        self.can_attack = None
        self.can_move = None
        self.military = None
        self.structure = None
        self.recallable = False

        self.lumber_cost = None
        self.gold_cost = None
        self.oil_cost = None

        self.food_production = None
        self.food_consumption = None

        self.width = None
        self.height = None

        self.name = None
        self.state = self.state_manager.despawned_state
        self.state_queue = deque()

        # Spawning attributes
        self.spawn_duration = None
        self.spawn_timer = None
        self.spawn_tile_id = Constants.Tile.Nothing
        self.built_by_id = Constants.Unit.Nothing

        # Harvesting attributes
        self.harvest_interval = .5 * Config.TICK_MULTIPLIER
        self.harvest_timer = None
        self.harvest_iterator = None
        self.harvest_target_id = Constants.Unit.Nothing

        # Building attributes
        self.combat_target_id = Constants.Unit.Nothing
        self.combat_timer = 1000
        self.combat_interval = 1 * Config.TICK_MULTIPLIER

        # Walking
        self.walking_timer = 0
        self.walking_path = []
        self.walking_goal_id = Constants.Tile.Nothing
        self.walking_interval = .1 * Config.TICK_MULTIPLIER



    def update(self):
        self.state.update(self)

    def unit_area(self):
        if not self.d['x'] or not self.d['y']:
            return []
        map_tiles_x = [x for x in range(self.d['x'], self.d['x'] + self.width)]
        map_tiles_y = [y for y in range(self.d['y'], self.d['y'] + self.height)]
        tiles = list(itertools.product(map_tiles_x, map_tiles_y))
        return tiles

    def set_position(self, x, y):
        prev_x = self.d['x'] if self.d['x'] else x
        prev_y = self.d['y'] if self.d['y'] else y

        # If unit already have a position
        if self.d['x'] and self.d['y']:
            self.clear_position_matrix()

        # Set new position
        self.d['x'] = x
        self.d['y'] = y

        # Update Matrix
        self.update_position_matrix()

        # Set direction of step
        self.set_direction(self.d['x'] - prev_x, self.d['y'] - prev_y)


    def set_direction(self, dx, dy):
        data = {
            (1, 0): const.Unit.DIRECTION_RIGHT,
            (-1, 0): const.Unit.DIRECTION_LEFT,
            (0, 1): const.Unit.DIRECTION_DOWN,
            (0, 0): const.Unit.DIRECTION_DOWN,
            (0, -1): const.Unit.DIRECTION_UP,
            (1, 1): const.Unit.DIRECTION_DOWN_RIGHT,
            (-1, 1): const.Unit.DIRECTION_DOWN_LEFT,
            (-1, -1): const.Unit.DIRECTION_UP_LEFT,
            (1, -1): const.Unit.DIRECTION_UP_RIGHT
        }
        try:
            self.d['direction'] = data[(dx, dy)]
        except:
            self.d['direction'] = data[(0, 0)]     # TODO!

    def clear_position_matrix(self):
        for x, y in self.unit_area():
            del self.data['unit'][x, y]
            del self.data['unit_pid'][x, y]

    def update_position_matrix(self):
        for x, y in self.unit_area():
            self.game.data['unit'][x, y] = self.unit_id
            self.game.data['unit_pid'][x, y] = self.player.id

    def add_to_game(self):
        self.game.units[self.unit_id] = self
        self.player.units.append(self.unit_id)

    def remove_from_game(self):
        del self.game.units[self.unit_id]
        self.player.units.remove(self.unit_id)

    def spawn(self, spawn_tile, build_timer_value):
        self.spawn_timer = build_timer_value
        self.spawn_tile_id = spawn_tile.id
        self.transition_state_next(self.state_manager.spawn_state)
        self.enqueue_state(self.state_manager.idle_state)

    def transition_state_next(self, next_state):
        self.state.end(self)
        self.state = next_state
        self.state.init(self)

    def transition_state(self):
        if not self.state_queue:
            # Has state
            next_state = self.state_manager.idle_state
            self.transition_state_next(next_state)
            return

        next_state = self.state_queue.pop()
        self.state.end(self)
        self.state = next_state
        return



    def enqueue_state(self, state):
        self.state_queue.append(state)

    def despawn(self):
        self.clear_position_matrix()

        self.d['x'] = None
        self.d['y'] = None

        self.player.consumed_food -= self.food_cost
        self.player.food -= self.food

        self.d['spawned'] = False

    def generate_path(self, x, y):
        path = a_star_search(self.game, self.d['x'], self.d['y'], x, y)

        if path:
            path.pop()

        return path

    def move(self, x, y):
        # Cannot move if building
        if self.structure == const.Unit.TYPE_STRUCTURE or self.speed <= 0:
            return False

        path = [(x, y)] if self.distance(x, y ) == 1 else self.generate_path(x, y)

        self.state.transition(const.State.ID_Walking, {
            'path': path,
            'path_goal': (x, y)
        })

    def void(self):
        """
        A "do nothing" action
        :return:
        """
        return True

    def distance(self, x, y, dim=0):
        d = math.hypot(self.d['x'] - (x + dim), self.d['y'] - (y + dim))
        return int(d) - dim

    def _crossover(self, x, y):
        dx1 = self.d['x'] - x
        dy1 = self.d['y'] - y

        cross = abs(dx1) + abs(dy1)
        return cross*0.001

    def shortest_distance(self, tiles):
        tiles.sort(key=lambda x: self.distance(x[0], x[1]) + self._crossover(x[0], x[1]), reverse=True)
        return tiles.pop()

    def _dimension(self, clazz = None):

        width = self.width
        height = self.height
        if clazz:
            width = clazz.width
            height = clazz.height

        mid_x = math.floor(width / 2)
        mid_y = math.floor(height / 2)

        return int((mid_x + mid_y) / 2)

    def center_tile(self):
        """
        Since self.d['x'] and self.d['y'] are the upper left corner,
        on structures which are 3x3 we need to determine center
        :return:
        """
        mid_x = math.floor(self.width / 2)
        mid_y = math.floor(self.height / 2)
        return self.d['x'] - mid_x, self.d['y'] - mid_y

    def harvest(self, x, y):
        if not self.can_harvest:
            return

        distance = self.distance(x, y)
        if distance > 1:
            tiles = self.game.AdjacentMap.adjacent_walkable(self.game, x, y, 1)

            if not tiles:
                return
            tile = self.shortest_distance(tiles)

            if self.state.has_next_state(const.State.ID_Harvesting):
                self.state.clear_next()

            self.state.add_next(const.State.ID_Harvesting, {
                'harvest_target': (x, y)
            }, 1)
            self.move(*tile)
        else:
            self.state.transition(const.State.ID_Harvesting, {
                'harvest_target': (x, y)
            })

    def closest_recall_building(self):
        units = self.game.units
        recall_buildings = [units[x] for x in self.player.units if units[x].id in [
            const.Unit.ID_TOWN_HALL,
            const.Unit.ID_KEEP,
            const.Unit.ID_CASTLE
        ]]

        recall_buildings.sort(key=lambda x: self.distance(x.d['x'], x.d['y']), reverse=True)
        try:
            recall_building = recall_buildings.pop()
        except:
            return None

        return recall_building

    def is_walking(self):
        return self.state.type == const.State.ID_Walking

    def is_worker(self):
        return self.can_harvest

    def get_damage(self, unit):
        # http://classic.battle.net/war2/basic/combat.shtml
        my_damage = (random.randint(self.damage_min, self.damage_max) - unit.armor) + self.damage_piercing
        return random.randint(math.floor(my_damage * .50), my_damage)

    def afflict_damage(self, damage):
        self.d['health'] = max(0, self.d['health'] - damage)

        if self.d['health'] <= 0:
            self.state.transition(const.State.ID_Dead)

            # In addition check for @see issue #1
            self.player.calculate_defeat()
            self.game.calculate_winner()

    def is_dead(self):
        return self.state.type == const.State.ID_Dead

    def attack(self, x, y):

        # Units with no minimum damage cannot attack
        # Warning! This makes unit not move to attack target, if they do not have damage!
        if self.damage_min <= 0:
            return

        attack_target = self.game.get_unit(x, y)
        distance = self.distance(x, y)
        if distance > 1:

            tiles = self.AdjacentMap.adjacent_walkable(
                self.game,
                attack_target.d['x'] + attack_target.dimension,
                attack_target.d['y'] + attack_target.dimension,
                attack_target.dimension + 1
            )

            if not tiles:
                return
            tile = self.shortest_distance(tiles)

            self.state.add_next(const.State.ID_Combat, {
                'attack_target': attack_target.unit_id
            }, 2)

            self.move(*tile)
        else:
            self.state.transition(const.State.ID_Combat, {
                'attack_target': attack_target.unit_id
            })

    def right_click(self, x, y):
        if not self.d['x'] or not self.d['y'] or self.state.type == const.State.ID_Building:
            return
        if self.game.Map.is_attackable(self, x, y):
            self.states['q'].clear()
            self.attack(x, y)
        elif self.structure == const.Unit.TYPE_UNIT and self.game.Map.is_harvestable_tile(self, x, y):
            self.states['q'].clear()
            self.harvest(x, y)
        elif self.game.Map.is_walkable_tile(self, x, y):
            self.states['q'].clear()
            self.move(x, y)

    def can_build(self, entity):

        if self.player.consumed_food + entity.food_cost > self.player.food:
            #print("Not enough food!")
            return False

        if self.player.gold < entity.cost_gold or self.player.lumber < entity.cost_lumber:
            #print("Cannot afford this unit")
            return False

        if self.state.type != const.State.ID_Idle and self.state.type != const.State.ID_Walking:
            #print("Wrong state")
            return False

        return True

    def can_build_here(self, subject):
        """
        This determines weither this unit can build a unit at current location
        :param subject:
        :param x:
        :param y:
        :return:
        """

        if self.structure == const.Unit.TYPE_STRUCTURE:
            adjacent_tiles = self.AdjacentMap.adjacent_walkable(
                self.game,
                self.d['x'] + self.dimension,
                self.d['y'] + self.dimension,
                self.dimension + 1
            )
            return len(adjacent_tiles) > 0
        else:

            # A unit builds a structure
            adjacent_tiles = self.AdjacentMap.adjacent_walkable(self.game, self.d['x'], self.d['y'], Unit._dimension(subject))
            adjacent_tiles.append((self.d['x'], self.d['y']))
            subject_area = list(
                itertools.product(
                    list(range(self.d['x'] - Unit._dimension(subject), self.d['x'] - Unit._dimension(subject) + subject.height)),
                    list(range(self.d['y'] - Unit._dimension(subject), self.d['y'] - Unit._dimension(subject) + subject.width))
                ))
            common = list(set(adjacent_tiles).intersection(set(subject_area)))
            return len(common) == len(subject_area)

    def build(self, idx):

        if self.state.id != Constants.State.Idle:
            return False

        if idx < 0 or idx >= len(self.build_inventory):
            return False

        build_unit_idx = self.build_inventory[idx]
        new_unit = UnitManager.construct_unit(build_unit_idx, self.player)


        if not self.can_build(entity_class) and not force_allow:
            return False

        if self.can_build_here(entity_class):
            # All tests passed, unit can be built
            # Subtract resources
            self.player.gold -= entity_class.cost_gold
            self.player.lumber -= entity_class.cost_lumber

            self.player.statistics['unit_count'] += 1

            self.state.transition(const.State.ID_Building, {
                'target_id': entity_class.id,
                'build_duration': entity_class.build_duration,
                'build_timer': 0
            })

            return True
        else:
            # TODO feedback could not build
            return False
            pass
